File: src/install/PackageManager/UpdateRequest.zig
name: string = "",
name_hash: PackageNameHash = 0,
version: Dependency.Version = .{},
version_buf: []const u8 = "",
package_id: PackageID = invalid_package_id,
is_aliased: bool = false,
failed: bool = false,
// This must be cloned to handle when the AST store resets
e_string: ?*JSAst.E.String = null,

pub const Array = std.ArrayListUnmanaged(UpdateRequest);

pub inline fn matches(this: PackageManager.UpdateRequest, dependency: Dependency, string_buf: []const u8) bool {
    return this.name_hash == if (this.name.len == 0)
        String.Builder.stringHash(dependency.version.literal.slice(string_buf))
    else
        dependency.name_hash;
}

pub fn getName(this: *const UpdateRequest) string {
    return if (this.is_aliased)
        this.name
    else
        this.version.literal.slice(this.version_buf);
}

/// If `this.package_id` is not `invalid_package_id`, it must be less than `lockfile.packages.len`.
pub fn getNameInLockfile(this: *const UpdateRequest, lockfile: *const Lockfile) ?string {
    return if (this.package_id == invalid_package_id)
        null
    else
        lockfile.packages.items(.name)[this.package_id].slice(this.version_buf);
}

/// It is incorrect to call this function before Lockfile.cleanWithLogger() because
/// resolved_name should be populated if possible.
///
/// `this` needs to be a pointer! If `this` is a copy and the name returned from
/// resolved_name is inlined, you will return a pointer to stack memory.
pub fn getResolvedName(this: *const UpdateRequest, lockfile: *const Lockfile) string {
    return if (this.is_aliased)
        this.name
    else if (this.getNameInLockfile(lockfile)) |name|
        name
    else
        this.version.literal.slice(this.version_buf);
}

pub fn fromJS(globalThis: *jsc.JSGlobalObject, input: jsc.JSValue) bun.JSError!jsc.JSValue {
    var arena = std.heap.ArenaAllocator.init(bun.default_allocator);
    defer arena.deinit();
    var stack = std.heap.stackFallback(1024, arena.allocator());
    const allocator = stack.get();
    var all_positionals = std.array_list.Managed([]const u8).init(allocator);

    var log = logger.Log.init(allocator);

    if (input.isString()) {
        var input_str = try input.toSliceCloneWithAllocator(
            globalThis,
            allocator,
        );
        if (input_str.len > 0)
            try all_positionals.append(input_str.slice());
    } else if (input.isArray()) {
        var iter = try input.arrayIterator(globalThis);
        while (try iter.next()) |item| {
            const slice = try item.toSliceCloneWithAllocator(globalThis, allocator);
            if (slice.len == 0) continue;
            try all_positionals.append(slice.slice());
        }
    } else {
        return .js_undefined;
    }

    if (all_positionals.items.len == 0) {
        return .js_undefined;
    }

    var array = Array{};

    const update_requests = parseWithError(allocator, null, &log, all_positionals.items, &array, .add, false) catch {
        return globalThis.throwValue(try log.toJS(globalThis, bun.default_allocator, "Failed to parse dependencies"));
    };
    if (update_requests.len == 0) return .js_undefined;

    if (log.msgs.items.len > 0) {
        return globalThis.throwValue(try log.toJS(globalThis, bun.default_allocator, "Failed to parse dependencies"));
    }

    if (update_requests[0].failed) {
        return globalThis.throw("Failed to parse dependencies", .{});
    }

    var object = jsc.JSValue.createEmptyObject(globalThis, 2);
    var name_str = bun.String.init(update_requests[0].name);
    object.put(globalThis, "name", try name_str.transferToJS(globalThis));
    object.put(globalThis, "version", try update_requests[0].version.toJS(update_requests[0].version_buf, globalThis));
    return object;
}

pub fn parse(
    allocator: std.mem.Allocator,
    pm: ?*PackageManager,
    log: *logger.Log,
    positionals: []const string,
    update_requests: *Array,
    subcommand: Subcommand,
) []UpdateRequest {
    return parseWithError(allocator, pm, log, positionals, update_requests, subcommand, true) catch Global.crash();
}

fn parseWithError(
    allocator: std.mem.Allocator,
    pm: ?*PackageManager,
    log: *logger.Log,
    positionals: []const string,
    update_requests: *Array,
    subcommand: Subcommand,
    fatal: bool,
) ![]UpdateRequest {
    // first one is always either:
    // add
    // remove
    outer: for (positionals) |positional| {
        var input: []u8 = bun.handleOom(bun.default_allocator.dupe(u8, std.mem.trim(u8, positional, " \n\r\t")));
        {
            var temp: [2048]u8 = undefined;
            const len = std.mem.replace(u8, input, "\\\\", "/", &temp);
            bun.path.platformToPosixInPlace(u8, &temp);
            const input2 = temp[0 .. input.len - len];
            @memcpy(input[0..input2.len], input2);
            input.len = input2.len;
        }
        switch (subcommand) {
            .link, .unlink => if (!strings.hasPrefixComptime(input, "link:")) {
                input = std.fmt.allocPrint(allocator, "{0s}@link:{0s}", .{input}) catch unreachable;
            },
            else => {},
        }

        var value = input;
        var alias: ?string = null;
        if (!Dependency.isTarball(input) and strings.isNPMPackageName(input)) {
            alias = input;
            value = input[input.len..];
        } else if (input.len > 1) {
            if (strings.indexOfChar(input[1..], '@')) |at| {
                const name = input[0 .. at + 1];
                if (strings.isNPMPackageName(name)) {
                    alias = name;
                    value = input[at + 2 ..];
                }
            }
        }

        const placeholder = String.from("@@@");
        var version = Dependency.parseWithOptionalTag(
            allocator,
            if (alias) |name| String.init(input, name) else placeholder,
            if (alias) |name| String.Builder.stringHash(name) else null,
            value,
            null,
            &SlicedString.init(input, value),
            log,
            pm,
        ) orelse {
            if (fatal) {
                Output.errGeneric("unrecognised dependency format: {s}", .{
                    positional,
                });
            } else {
                log.addErrorFmt(null, logger.Loc.Empty, allocator, "unrecognised dependency format: {s}", .{
                    positional,
                }) catch |err| bun.handleOom(err);
            }

            return error.UnrecognizedDependencyFormat;
        };
        if (alias != null and version.tag == .git) {
            if (Dependency.parseWithOptionalTag(
                allocator,
                placeholder,
                null,
                input,
                null,
                &SlicedString.init(input, input),
                log,
                pm,
            )) |ver| {
                alias = null;
                version = ver;
            }
        }
        if (switch (version.tag) {
            .dist_tag => version.value.dist_tag.name.eql(placeholder, input, input),
            .npm => version.value.npm.name.eql(placeholder, input, input),
            else => false,
        }) {
            if (fatal) {
                Output.errGeneric("unrecognised dependency format: {s}", .{
                    positional,
                });
            } else {
                log.addErrorFmt(null, logger.Loc.Empty, allocator, "unrecognised dependency format: {s}", .{
                    positional,
                }) catch |err| bun.handleOom(err);
            }

            return error.UnrecognizedDependencyFormat;
        }

        var request = UpdateRequest{
            .version = version,
            .version_buf = input,
        };
        if (alias) |name| {
            request.is_aliased = true;
            request.name = allocator.dupe(u8, name) catch unreachable;
            request.name_hash = String.Builder.stringHash(name);
        } else if (version.tag == .github and version.value.github.committish.isEmpty()) {
            request.name_hash = String.Builder.stringHash(version.literal.slice(input));
        } else {
            request.name_hash = String.Builder.stringHash(version.literal.slice(input));
        }

        for (update_requests.items) |*prev| {
            if (prev.name_hash == request.name_hash and request.name.len == prev.name.len) continue :outer;
        }
        bun.handleOom(update_requests.append(allocator, request));
    }

    return update_requests.items;
}

pub const CommandLineArguments = PackageManager.CommandLineArguments;
pub const Options = PackageManager.Options;
pub const PackageInstaller = PackageManager.PackageInstaller;
pub const PackageJSONEditor = PackageManager.PackageJSONEditor;
pub const Subcommand = PackageManager.Subcommand;

const string = []const u8;

const std = @import("std");

const bun = @import("bun");
const Global = bun.Global;
const JSAst = bun.ast;
const Output = bun.Output;
const default_allocator = bun.default_allocator;
const jsc = bun.jsc;
const logger = bun.logger;
const strings = bun.strings;

const Semver = bun.Semver;
const SlicedString = Semver.SlicedString;
const String = Semver.String;

const Dependency = bun.install.Dependency;
const Lockfile = bun.install.Lockfile;
const PackageID = bun.install.PackageID;
const PackageNameHash = bun.install.PackageNameHash;
const invalid_package_id = bun.install.invalid_package_id;

const PackageManager = bun.install.PackageManager;
const UpdateRequest = PackageManager.UpdateRequest;

File: src/install/PackageManager/UpdateRequest.rs
use crate::lockfile::package::PackageColumns as _;
use std::io::Write as _;

use bun_ast::{Loc, Log};
use bun_core::strings;
use bun_core::{Global, Output};
use bun_js_parser as js_ast;
use bun_semver::{SlicedString, String as SemverString, string::Builder as StringBuilder};

use bun_install::dependency::{self, DependencyExt as _};
use bun_install::{
    Dependency, INVALID_PACKAGE_ID, Lockfile, PackageID, PackageManager, PackageNameHash,
};
// `lockfile.packages.items_name()` is provided by an extension trait on
// `MultiArrayList<Package>` (Zig: `lockfile.packages.items(.name)`).
pub struct UpdateRequest {
    // TODO(port): lifetime — Zig leaks these (no deinit); using &'static for now
    pub name: &'static [u8],
    pub name_hash: PackageNameHash,
    pub version: dependency::Version,
    /// Backing buffer for `version.literal` (and friends) — either a leaked
    /// CLI positional (truly process-lifetime) or the active lockfile's
    /// `buffers.string_bytes`. Stored as a raw fat pointer because the
    /// lockfile buffer's lifetime cannot be expressed as `'static` without UB
    /// lifetime extension (PORTING.md §Forbidden patterns), and threading a
    /// real `<'a>` through every `&mut [UpdateRequest]` in the install
    /// pipeline is the Phase-B reshape. ARENA-class field per the PORTING.md
    /// type map: `[]const u8` struct-field, never freed, points into a buffer
    /// owned elsewhere → `RawSlice<u8>` (centralises the outlives-holder
    /// invariant; see `version_buf()`).
    pub version_buf: bun_ptr::RawSlice<u8>,
    pub package_id: PackageID,
    pub is_aliased: bool,
    pub failed: bool,
    /// This must be cloned to handle when the AST store resets
    // TODO(port): lifetime — ARENA-owned (AST Expr.Data store); raw ptr per LIFETIMES.tsv
    pub e_string: Option<*mut js_ast::E::String>,
}

impl Default for UpdateRequest {
    fn default() -> Self {
        Self {
            name: b"",
            name_hash: 0,
            version: dependency::Version::default(),
            version_buf: bun_ptr::RawSlice::EMPTY,
            package_id: INVALID_PACKAGE_ID,
            is_aliased: false,
            failed: false,
            e_string: None,
        }
    }
}

pub type Array = Vec<UpdateRequest>;

impl UpdateRequest {
    /// Borrow the backing string buffer.
    ///
    /// SAFETY for callers: the buffer this points into (leaked CLI input, or
    /// `lockfile.buffers.string_bytes` after `clean_with_logger`) must outlive
    /// the returned slice and must not be reallocated while the borrow is
    /// live. Both invariants hold on every call path today — `string_bytes`
    /// is finalized before assignment in `clean_with_logger`, and the lockfile
    /// is threaded alongside `updates` everywhere they are read.
    #[inline]
    pub fn version_buf(&self) -> &[u8] {
        // See fn doc. `RawSlice` encapsulates the deref under the
        // outlives-holder invariant; `Default` seeds it as `EMPTY` and every
        // assignment is `RawSlice::new(&[u8])`.
        self.version_buf.slice()
    }

    #[inline]
    pub fn matches(&self, dependency: &Dependency, string_buf: &[u8]) -> bool {
        self.name_hash
            == if self.name.is_empty() {
                StringBuilder::string_hash(dependency.version.literal.slice(string_buf))
            } else {
                dependency.name_hash
            }
    }

    pub fn get_name(&self) -> &[u8] {
        if self.is_aliased {
            self.name
        } else {
            self.version.literal.slice(self.version_buf())
        }
    }

    /// If `self.package_id` is not `invalid_package_id`, it must be less than `lockfile.packages.len`.
    pub fn get_name_in_lockfile<'a>(&'a self, lockfile: &'a Lockfile) -> Option<&'a [u8]> {
        if self.package_id == INVALID_PACKAGE_ID {
            None
        } else {
            Some(lockfile.packages.items_name()[self.package_id as usize].slice(self.version_buf()))
        }
    }

    /// It is incorrect to call this function before Lockfile.cleanWithLogger() because
    /// resolved_name should be populated if possible.
    ///
    /// `self` needs to be a pointer! If `self` is a copy and the name returned from
    /// resolved_name is inlined, you will return a pointer to stack memory.
    pub fn get_resolved_name<'a>(&'a self, lockfile: &'a Lockfile) -> &'a [u8] {
        if self.is_aliased {
            self.name
        } else if let Some(name) = self.get_name_in_lockfile(lockfile) {
            name
        } else {
            self.version.literal.slice(self.version_buf())
        }
    }

    // NOTE: `pub const fromJS = @import("../../install_jsc/update_request_jsc.zig").fromJS;`
    // deleted — in Rust, `from_js` lives on an extension trait in the `*_jsc` crate.

    pub fn parse<'a>(
        pm: Option<&mut PackageManager>,
        log: &mut Log,
        positionals: &[&[u8]],
        update_requests: &'a mut Array,
        subcommand: Subcommand,
    ) -> &'a mut [UpdateRequest] {
        Self::parse_with_error(pm, log, positionals, update_requests, subcommand, true)
            .unwrap_or_else(|_| Global::crash())
    }

    // TODO(port): narrow error set — only `UnrecognizedDependencyFormat` is returned
    pub fn parse_with_error<'a>(
        mut pm: Option<&mut PackageManager>,
        log: &mut Log,
        positionals: &[&[u8]],
        update_requests: &'a mut Array,
        subcommand: Subcommand,
        fatal: bool,
    ) -> Result<&'a mut [UpdateRequest], bun_core::Error> {
        // first one is always either:
        // add
        // remove
        'outer: for positional in positionals {
            let mut input: Vec<u8> = strings::trim(positional, b" \n\r\t").to_vec();
            {
                // Replacing "\\\\" (2 bytes) with "/" (1 byte) never grows the string, so a
                // buffer of `input.len` bytes is always sufficient. Previously this was a
                // fixed `[2048]u8` stack array which overflowed for longer positionals.
                let mut temp = vec![0u8; input.len()];
                // std.mem.replace(u8, input, "\\\\", "/", temp) — returns replacement count
                let len = strings::replace(&input, b"\\\\", b"/", &mut temp);
                let new_len = input.len() - len;
                let input2 = &mut temp[..new_len];
                bun_paths::resolve_path::platform_to_posix_in_place(input2);
                input[..new_len].copy_from_slice(input2);
                input.truncate(new_len);
            }
            match subcommand {
                Subcommand::Link | Subcommand::Unlink => {
                    if !input.starts_with(b"link:") {
                        let mut buf = Vec::with_capacity(input.len() * 2 + 6);
                        write!(&mut buf, "{0}@link:{0}", bstr::BStr::new(&input))
                            .expect("unreachable");
                        input = buf;
                    }
                }
                _ => {}
            }

            // PORT NOTE: reshaped for borrowck — leak `input` now so sub-slices are &'static.
            // Zig: `bun.default_allocator.dupe(u8, ..)` with no matching free; these live for
            // the CLI invocation. `version_buf` is later reassigned to point at lockfile
            // buffers (lockfile.rs), so the field is a raw `*const [u8]` (ARENA-class per
            // PORTING.md type map) rather than `Box<[u8]>`.
            let input: &'static [u8] = input.leak();

            let mut value: &'static [u8] = input;
            let mut alias: Option<&'static [u8]> = None;
            if !Dependency::is_tarball(input) && strings::is_npm_package_name(input) {
                alias = Some(input);
                value = &input[input.len()..];
            } else if input.len() > 1 {
                if let Some(at) = strings::index_of_char(&input[1..], b'@') {
                    let name = &input[0..at as usize + 1];
                    if strings::is_npm_package_name(name) {
                        alias = Some(name);
                        value = &input[at as usize + 2..];
                    }
                }
            }

            let placeholder = SemverString::from(b"@@@");
            let Some(mut version) = Dependency::parse_with_optional_tag(
                if let Some(name) = alias {
                    SemverString::init(input, name)
                } else {
                    placeholder
                },
                alias.map(|name| StringBuilder::string_hash(name)),
                value,
                None,
                &mut SlicedString::init(input, value),
                Some(&mut *log),
                pm.as_deref_mut(),
            ) else {
                if fatal {
                    Output::err_generic(
                        "unrecognised dependency format: {}",
                        format_args!("{}", bstr::BStr::new(positional)),
                    );
                } else {
                    log.add_error_fmt(
                        None,
                        Loc::EMPTY,
                        format_args!(
                            "unrecognised dependency format: {}",
                            bstr::BStr::new(positional)
                        ),
                    );
                }

                return Err(bun_core::err!("UnrecognizedDependencyFormat"));
            };
            // TODO(port): Dependency.Version tag/value layout — Zig uses separate .tag + .value union
            if alias.is_some() && version.tag == dependency::version::Tag::Git {
                if let Some(ver) = Dependency::parse_with_optional_tag(
                    placeholder,
                    None,
                    input,
                    None,
                    &mut SlicedString::init(input, input),
                    Some(&mut *log),
                    pm.as_deref_mut(),
                ) {
                    alias = None;
                    version = ver;
                }
            }
            if match version.tag {
                dependency::version::Tag::DistTag => {
                    version.dist_tag().name.eql(placeholder, input, input)
                }
                dependency::version::Tag::Npm => version.npm().name.eql(placeholder, input, input),
                _ => false,
            } {
                if fatal {
                    Output::err_generic(
                        "unrecognised dependency format: {}",
                        format_args!("{}", bstr::BStr::new(positional)),
                    );
                } else {
                    log.add_error_fmt(
                        None,
                        Loc::EMPTY,
                        format_args!(
                            "unrecognised dependency format: {}",
                            bstr::BStr::new(positional)
                        ),
                    );
                }

                return Err(bun_core::err!("UnrecognizedDependencyFormat"));
            }

            let mut request = UpdateRequest {
                version,
                version_buf: bun_ptr::RawSlice::new(input),
                ..UpdateRequest::default()
            };
            if let Some(name) = alias {
                request.is_aliased = true;
                // Zig: `allocator.dupe(u8, name) catch unreachable` — never freed (CLI lifetime).
                request.name = name.to_vec().leak();
                request.name_hash = StringBuilder::string_hash(name);
            } else if request.version.tag == dependency::version::Tag::Github
                && request.version.github().committish.is_empty()
            {
                request.name_hash =
                    StringBuilder::string_hash(request.version.literal.slice(input));
            } else {
                request.name_hash =
                    StringBuilder::string_hash(request.version.literal.slice(input));
            }

            for prev in update_requests.iter() {
                if prev.name_hash == request.name_hash && request.name.len() == prev.name.len() {
                    continue 'outer;
                }
            }
            update_requests.push(request);
        }

        Ok(update_requests.as_mut_slice())
    }
}

pub use super::Subcommand;
pub use bun_install::package_manager::Options;
pub use bun_install::package_manager::command_line_arguments as CommandLineArguments;

// ported from: src/install/PackageManager/UpdateRequest.zig
